src/main/java/com/aws/greengrass/clientdevices/auth/util/WildcardTrie.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.clientdevices.auth.util;
+
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Supplier;
+
+public class WildcardTrie {
+    private static final String GLOB_WILDCARD = "*";
+    private static final String SINGLE_CHAR_WILDCARD = "?";
+
+    private final Map<String, WildcardTrie> children = new DefaultHashMap<>(WildcardTrie::new);
+
+    private boolean isTerminal;
+    private boolean isGlobWildcard;
+    private boolean isSingleCharWildcard;
+
+    public void add(String subject) {
+        add(subject, true);
+    }
+
+    private WildcardTrie add(String subject, boolean isTerminal) {
+        if (subject == null || subject.isEmpty()) {
+            this.isTerminal |= isTerminal;
+            return this;
+        }
+        StringBuilder currPrefix = new StringBuilder(subject.length());
+        for (int i = 0; i < subject.length(); i++) {
+            char c = subject.charAt(i);
+            if (c == GLOB_WILDCARD.charAt(0)) {
+                return addGlobWildcard(subject, currPrefix.toString(), isTerminal);
+            }
+            if (c == SINGLE_CHAR_WILDCARD.charAt(0)) {
+                return addSingleCharWildcard(subject, currPrefix.toString(), isTerminal);
+            }
+            currPrefix.append(c);
+        }
+        WildcardTrie node = children.get(currPrefix.toString());
+        node.isTerminal |= isTerminal;
+        return node;
+    }
+
+    private WildcardTrie addGlobWildcard(String subject, String currPrefix, boolean isTerminal) {
+        WildcardTrie node = this;
+        node = node.add(currPrefix, false);
+        node = node.children.get(GLOB_WILDCARD);
+        node.isGlobWildcard = true;
+        // wildcard at end of subject is terminal
+        if (subject.length() - currPrefix.length() == 1) {
+            node.isTerminal = isTerminal;
+            return node;
+        }
+        return node.add(subject.substring(currPrefix.length() + 2), true);
+    }
+
+    private WildcardTrie addSingleCharWildcard(String subject, String currPrefix, boolean isTerminal) {
+        WildcardTrie node = this;
+        node = node.add(currPrefix, false);
+        node = node.children.get(SINGLE_CHAR_WILDCARD);
+        node.isSingleCharWildcard = true;
+        // wildcard at end of subject is terminal
+        if (subject.length() - currPrefix.length() == 1) {
+            node.isTerminal = isTerminal;
+            return node;
+        }
+        return node.add(subject.substring(currPrefix.length() + 1), true);
+    }
+
+    public boolean matches(String s) {
+        return matches(s, true);
+    }
+
+    public boolean matches(String s, boolean matchSingleCharWildcard) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 21. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 123-132 into a separate method.